### ISSUE TITLE
[REVIEW]: DataDepsGenerators.jl: making reusing data easy by automatically generating DataDeps.jl registration code

### ISSUE THREAD

**Issue by whedon** [openjournals/joss-reviews#921]
**Submitting author:** @oxinabox (<a href="http://orcid.org/0000-0003-1386-1646">Lyndon White</a>)
**Repository:** <a href="https://github.com/oxinabox/DataDepsGenerators.jl/" target ="_blank">https://github.com/oxinabox/DataDepsGenerators.jl/</a>
**Version:** v0.4.0
**Editor:** @arfon
**Reviewer:** @ninjin
**Archive:** <a href="https://doi.org/10.5281/zenodo.1478705" target="_blank">10.5281/zenodo.1478705</a>

## Status

[![status](http://joss.theoj.org/papers/f52340014957dc0e74d5935162221c29/status.svg)](http://joss.theoj.org/papers/f52340014957dc0e74d5935162221c29)

Status badge code:

```
HTML: <a href="http://joss.theoj.org/papers/f52340014957dc0e74d5935162221c29"><img src="http://joss.theoj.org/papers/f52340014957dc0e74d5935162221c29/status.svg"></a>
Markdown: [![status](http://joss.theoj.org/papers/f52340014957dc0e74d5935162221c29/status.svg)](http://joss.theoj.org/papers/f52340014957dc0e74d5935162221c29)
```
**Reviewers and authors**:

Please avoid lengthy details of difficulties in the review thread. Instead, please create a new issue in the <a href="https://github.com/oxinabox/DataDepsGenerators.jl/" target="_blank">target repository</a> and link to those issues (especially acceptance-blockers) in the review thread below. (For completists: if the target issue tracker is also on GitHub, linking the review thread in the issue or vice versa will create corresponding breadcrumb trails in the link target.)

## Reviewer instructions & questions

@ninjin, please carry out your review in this issue by updating the checklist below. If you cannot edit the checklist please:
1. Make sure you're logged in to your GitHub account
2. Be sure to accept the invite at this URL: https://github.com/openjournals/joss-reviews/invitations

The reviewer guidelines are available here: https://joss.theoj.org/about#reviewer_guidelines. Any questions/concerns please let @arfon know.

✨ **Please try and complete your review in the next two weeks** ✨ 

### Conflict of interest

- [x] As the reviewer I confirm that I have read the [JOSS conflict of interest policy](https://github.com/openjournals/joss/blob/master/COI.md) and that there are no conflicts of interest for me to review this work.

### Code of Conduct

- [x] I confirm that I read and will adhere to the [JOSS code of conduct](https://joss.theoj.org/about#code_of_conduct).

### General checks

- [x] **Repository:** Is the source code for this software available at the <a target="_blank" href="https://github.com/oxinabox/DataDepsGenerators.jl/">repository url</a>?
- [x] **License:** Does the repository contain a plain-text LICENSE file with the contents of an [OSI approved](https://opensource.org/licenses/alphabetical) software license?
- [x] **Version:** Does the release version given match the GitHub release (v0.4.0)?
- [x] **Authorship:** Has the submitting author (@oxinabox) made major contributions to the software? Does the full list of paper authors seem appropriate and complete?

### Functionality

- [x] **Installation:** Does installation proceed as outlined in the documentation?
- [x] **Functionality:** Have the functional claims of the software been confirmed?
- [x] **Performance:** If there are any performance claims of the software, have they been confirmed? (If there are no claims, please check off this item.)

### Documentation

- [x] **A statement of need:** Do the authors clearly state what problems the software is designed to solve and who the target audience is?
- [x] **Installation instructions:** Is there a clearly-stated list of dependencies? Ideally these should be handled with an automated package management solution.
- [x] **Example usage:** Do the authors include examples of how to use the software (ideally to solve real-world analysis problems).
- [x] **Functionality documentation:** Is the core functionality of the software documented to a satisfactory level (e.g., API method documentation)?
- [x] **Automated tests:** Are there automated tests or manual steps described so that the function of the software can be verified?
- [x] **Community guidelines:** Are there clear guidelines for third parties wishing to 1) Contribute to the software 2) Report issues or problems with the software 3) Seek support

### Software paper

- [x] **Authors:** Does the `paper.md` file include a list of authors with their affiliations?
- [x] **A statement of need:** Do the authors clearly state what problems the software is designed to solve and who the target audience is?
- [x] **References:** Do all archival references that should have a DOI list one (e.g., papers, datasets, software)?

**Comment by whedon** [comment]
[--> Check article proof :page_facing_up: <--](https://github.com/openjournals/joss-papers/blob/joss.00921/joss.00921/10.21105.joss.00921.pdf)

**Comment by ninjin** [comment]
Apologies for the delays, here is the current state of my review.

Formatting:

1. The citations look a bit odd to me, should “((Bezanson et al. 2014))” not be “(Bezanson et al. 2014)”?
2. A period is missing at the end of the third bullet in the first bullet list.
3. In the second to last paragraph there is what I think is a failed listing with hyphens following dots, this should probably be fixed.
4. I suspect that both of the cited works have DOIs assigned to them, Bezanson et al. (2014) I think was accepted by some IEEE journal. These DOIs should be added.

Content:

1. Are there not other similar packages for other programming languages? There is currently a bit of a lack of a wider context of open science, etc. and I think one or two sentences with a few cites would be very helpful for the readers.

Other than the above comments, the submitted paper reads very well.

Documentation:

1. No clear statement of need.
2. Misspellings: “consitantly”, “webbrowser”, “targetted”, etc. run a spell checker.
3. Awkward grammar in a number of places, just give it another read through.
4. While this is a small package, having at least a rudimentary `CONTRIBUTING.md` would be a good idea.

Functionality:

Currently blocked by [#62](https://github.com/oxinabox/DataDepsGenerators.jl/issues/62).

**Comment by oxinabox** [comment]
Thanks, I will attend to these in the next couple of days.

**Comment by oxinabox** [comment]
> I suspect that both of the cited works have DOIs assigned to them, Bezanson et al. (2014) I think was accepted by some IEEE journal. These DOIs should be added.

the DataDeps work is still under-review (and is taking a truly long time) and does not yet have a DOI.
I have searched [IEEE](https://ieeexplore.ieee.org/search/searchresult.jsp?newsearch=true&queryText=Bezanson) 
and I can not find 2014 julia paper there.
But it appears now the 2017 SIAM paper [is now the preferred citation for Julia, so I will update it to that](https://julialang.org/publications/),
that one does have a DOI.

**Comment by ninjin** [comment]
Your are indeed correct, it was a SIAM paper and not IEEE.

**Comment by oxinabox** [comment]
For my own records:
**Formatting** issues are fixed.

**Content**  issues are fixed, at least for R I added a section.
I don't know much about this space for Python.

@whedon generate pdf

**Documentation:**
Will be fixed once https://github.com/oxinabox/DataDepsGenerators.jl/pull/64
is merged

> No clear statement of need.

I took the first paragraph from the paper and added it to the docs,
since I think that sums it up pretty well.

> Misspellings: “consitantly”, “webbrowser”, “targetted”, etc. run a spell checker.

Done 

> Awkward grammar in a number of places, just give it another read through.

Done


> While this is a small package, having at least a rudimentary CONTRIBUTING.md would be a good idea

Done
--------------------------------------

Re Functionality being blocked.
That is now resolved for julia 0.7+/1.0  in the newly released DataDepsGenerators v0.5.0 
The DataDepsGenerators  0.4.0 release can be used on 0.6, but I agree with  @ninjin  it is better for him to review the julia 1.0 compatible version since that will become the version most people will use.

**Comment by arfon** [comment]
@whedon generate pdf

**Comment by whedon** [comment]
```
Attempting PDF compilation. Reticulating splines etc...
```

**Comment by whedon** [comment]
[ :point_right: Check article proof :page_facing_up: :point_left: ](https://github.com/openjournals/joss-papers/blob/joss.00921/joss.00921/10.21105.joss.00921.pdf)

**Comment by SebastinSanty** [comment]
Any updates on this?

**Comment by ninjin** [comment]
My apologies, will sort it out in the morning.

**Comment by ninjin** [comment]
I think GitHub is back online…

# Writing #

Nothing major, but some issues that should be addressed.

1. “leavages”
2. The CKAN entry has a bullet point split across two bullet points.
3. Same for JSON-LD.
4. “Julia” should always be capitalised.
5. “By generating code that the user may edit they may tweak the details as required.”, this one is a bit hard to parse, reformulate?
6. “If these APIs were to be access directly by a project, …”, not native, but “accessed”?
7. Maybe not use “things” in academic writing.
8. “the was proposed package (doidata) to acquire data based on a DOI.”, grammar.

# Code #

Pretty much works as intended, just some minor syntax issues in the README and some examples that were borked. Overall, I would say the code is good to go.

* [Syntax fix PR filed](https://github.com/oxinabox/DataDepsGenerators.jl/pull/66) (and merged)
* [Example issue](https://github.com/oxinabox/DataDepsGenerators.jl/issues/67)

**Comment by oxinabox** [comment]
@whedon generate pdf

Fixed the paper. thank you for the feedback.
(At some point I misread the guidance on the capitalization of julia, getting it literally backwards,
and thought it was  **julia** for the language, and **Julia** for the compiler.
But indeed it is the other way round (which makes much more sense since lowercase matches the executable name).

**Comment by whedon** [comment]
```
Attempting PDF compilation. Reticulating splines etc...
```

**Comment by whedon** [comment]
[ :point_right: Check article proof :page_facing_up: :point_left: ](https://github.com/openjournals/joss-papers/blob/joss.00921/joss.00921/10.21105.joss.00921.pdf)

**Comment by arfon** [comment]
Thanks @ninjin. @oxinabox - could you please do a final check of the spelling in your paper?

**Comment by oxinabox** [comment]
@arfon  I fixed up those spelling issues (though, I'll give it a 3rd check),
I was just waiting to fix up an issue @ninjin identified with one of our data providers,
https://github.com/oxinabox/DataDepsGenerators.jl/issues/67
Something in the DataONE API, or in the SSL client we use to access it changed in a breaking way in the last few months.
I am talking with the devs on both sides there.

@whedon generate pdf

**Comment by ninjin** [comment]
Fix “lavages” – [fun fact](https://en.wiktionary.org/wiki/lavage#English). Other than that is looks solid.

**Comment by oxinabox** [comment]
@whedon generate pdf

(sorry didn't do the 3rd check yet)

**Comment by whedon** [comment]
```
Attempting PDF compilation. Reticulating splines etc...
```

**Comment by whedon** [comment]
[ :point_right: Check article proof :page_facing_up: :point_left: ](https://github.com/openjournals/joss-papers/blob/joss.00921/joss.00921/10.21105.joss.00921.pdf)

**Comment by ninjin** [comment]
Looks good to me, @arfon do I sign off on the final version by closing this issue?

**Comment by arfon** [comment]
@oxinabox - At this point could you make an archive of the reviewed software in Zenodo/figshare/other service and update this thread with the DOI of the archive? I can then move forward with accepting the submission.

**Comment by arfon** [comment]
> Looks good to me, @arfon do I sign off on the final version by closing this issue?

Thanks @ninjin - I'll take it from here.

**Comment by oxinabox** [comment]
Done.

 DOI: 10.5281/zenodo.1478705

**Comment by arfon** [comment]
@whedon set 10.5281/zenodo.1478705 as archive

**Comment by whedon** [comment]
OK. <a href="https://doi.org/10.5281/zenodo.1478705" target="_blank">10.5281/zenodo.1478705</a> is the archive.

**Comment by arfon** [comment]
@whedon accept

**Comment by whedon** [comment]
```
Attempting dry run of processing paper acceptance...
```

**Comment by whedon** [comment]
Check final proof :point_right: https://github.com/openjournals/joss-papers/pull/57

If the paper PDF and Crossref deposit XML look good in https://github.com/openjournals/joss-papers/pull/57, then you can now move forward with accepting the submission by compiling again with the flag `deposit=true` e.g.
 ```
@whedon accept deposit=true
```

**Comment by arfon** [comment]
@whedon accept deposit=true

**Comment by whedon** [comment]
```
Doing it live! Attempting automated processing of paper acceptance...
```

**Comment by whedon** [comment]
🚨🚨🚨 **THIS IS NOT A DRILL, YOU HAVE JUST ACCEPTED A PAPER INTO JOSS!** 🚨🚨🚨

 Here's what you must now do:

0. Check final PDF and Crossref metadata that was deposited :point_right: https://github.com/openjournals/joss-papers/pull/58
1. Wait a couple of minutes to verify that the paper DOI resolves [https://doi.org/10.21105/joss.00921](https://doi.org/10.21105/joss.00921)
2. If everything looks good, then close this review issue.
3. Party like you just published a paper! 🎉🌈🦄💃👻🤘

 Any issues? notify your editorial technical team...

**Comment by arfon** [comment]
@ninjin - many thanks for your review here ✨ 

@oxinabox - your paper is now accepted into JOSS :zap: :rocket: :boom:

**Comment by whedon** [comment]
:tada::tada::tada: Congratulations on your paper acceptance! :tada::tada::tada:

If you would like to include a link to your paper from your README use the following code snippets:

```
Markdown:
[![DOI](http://joss.theoj.org/papers/10.21105/joss.00921/status.svg)](https://doi.org/10.21105/joss.00921)

HTML:
<a style="border-width:0" href="https://doi.org/10.21105/joss.00921">
  <img src="http://joss.theoj.org/papers/10.21105/joss.00921/status.svg" alt="DOI badge" >
</a>

reStructuredText:
.. image:: http://joss.theoj.org/papers/10.21105/joss.00921/status.svg
   :target: https://doi.org/10.21105/joss.00921
```
This is how it will look in your documentation:

[![DOI](http://joss.theoj.org/papers/10.21105/joss.00921/status.svg)](https://doi.org/10.21105/joss.00921)

**We need your help!**

Journal of Open Source Software is a community-run journal and relies upon volunteer effort. If you'd like to support us please consider doing either one (or both) of the the following:

- Volunteering to review for us sometime in the future. You can add your name to the reviewer list here: http://joss.theoj.org/reviewer-signup.html
- Making a small donation to support our running costs here: https://www.flipcause.com/secure/cause_pdetails/Mjk3Nzk=

**Comment by oxinabox** [comment]
@SebastinSanty  woot we did the thing.

@ninjin thanks again
